packages/metadata-editor/src/MetadataEditor.tsx

```diff
@@ -0,0 +1,253 @@
+ * limitations under the License.
+ */
+
+import { FormGroup } from '@blueprintjs/core';
+
+import { FrontendServices } from '@elyra/application';
+import { DropDown } from '@elyra/ui-components';
+
+import {
+  ReactWidget,
+  WidgetTracker,
+  showDialog,
+  Dialog
+} from '@jupyterlab/apputils';
+import { CodeEditor } from '@jupyterlab/codeeditor';
+import { InputGroup, Button } from '@jupyterlab/ui-components';
+
+import { Message } from '@lumino/messaging';
+
+import * as React from 'react';
+
+import { METADATA_EDITOR_ID } from './index';
+
+const ELYRA_METADATA_EDITOR_CLASS = 'elyra-metadataEditor';
+const DIRTY_CLASS = 'jp-mod-dirty';
+
+type FormItemType = 'TextInput' | 'DropDown' | 'Code';
+
+export type FormItem = {
+  value: any;
+  type: FormItemType;
+  label: string;
+  schemaField: string;
+};
+
+/**
+ * Metadata editor widget
+ */
+export class MetadataEditor extends ReactWidget {
+  metadata: FormItem[];
+  newFile: boolean;
+  updateSignal: any;
+  fileName: string;
+  editorFactory: CodeEditor.Factory;
+  editor: CodeEditor.IEditor;
+  namespace: string;
+  tracker: WidgetTracker;
+  dirty: boolean;
+
+  constructor(
+    metadata: FormItem[],
+    newFile: boolean,
+    updateSignal: any,
+    editorFactory: CodeEditor.Factory | null,
+    namespace: string,
+    tracker: WidgetTracker,
+    fileName?: string
+  ) {
+    super();
+    this.metadata = metadata;
+    this.editorFactory = editorFactory;
+    this.namespace = namespace;
+    this.newFile = newFile;
+    this.updateSignal = updateSignal;
+    this.handleTextInputChange = this.handleTextInputChange.bind(this);
+    this.handleDropdownChange = this.handleDropdownChange.bind(this);
+    this.tracker = tracker;
+    if (fileName) {
+      this.fileName = fileName;
+    }
+  }
+
+  onCloseRequest(msg: Message): void {
+    if (this.dirty) {
+      showDialog({
+        title: 'Close without saving?',
+        body: (
+          <p>
+            {' '}
+            {`Metadata "${this.fileName}" has unsaved changes, close without saving?`}{' '}
+          </p>
+        ),
+        buttons: [Dialog.cancelButton(), Dialog.okButton()]
+      }).then((response: any): void => {
+        if (response.button.accept) {
+          this.dispose();
+          super.onCloseRequest(msg);
+        }
+      });
+    } else {
+      this.dispose();
+      super.onCloseRequest(msg);
+    }
+  }
+
+  saveMetadata(): void {
+    const newMetadata: any = {
+      schema_name: 'code-snippet',
+      name: this.fileName,
+      display_name: this.getFormItem('Name').value,
+      metadata: {}
+    };
+
+    for (const field of this.metadata) {
+      if (field.type == 'TextInput') {
+        newMetadata.metadata[field.schemaField] = field.value;
+      } else if (field.type == 'DropDown') {
+        newMetadata.metadata[field.schemaField] = field.value.choice;
+      } else if (field.type == 'Code') {
+        newMetadata.metadata[field.schemaField] = field.value.split('\n');
```

Comment:
Same comment here.  If this is a generic base class, let the specific subclass deal with class-specific behaviors like 'Code'.